### PR TITLE
feat: add Node samples

### PR DESCRIPTION
**Pull request by ace-n**
~**DO NOT MERGE** until everyone's in sync on client library initialization (`DLP`/`dlp` shenanigans).~

~_Note: there are lots of merge conflicts due to whitespace changes. Once we've approved this PR, I can fix those locally and force-push the results to the `add-nodejs` branch._~

### CODE REVIEW

**Comment by ace-n** [comment]
@kurtisvg I was omitting Node samples for the non-Node specific stuff (to reduce maintenance load + visual clutter).

If you'd rather I change that, I'm happy to do so though!

**Comment by kurtisvg** [comment]
@ace-n I think it make's sense to have a different sample for each language, even if there aren't any major derivations.

File: content/en/_index.md
Comment (kurtisvg) [inline content/en/_index.md:140]: > Any declared function arguments should include a no-arg, main method

This shouldn't have arguments. However, I don't particularly care if it's called "main" since node doesn't require that - if it's more idiomatic to just call it "runSample" that's fine with me.

File: content/en/_index.md
Comment (kurtisvg) [inline content/en/_index.md:141]: I don't think we want region tags here - they should encapsulate the whole file of whatever this sample is.

File: content/en/_index.md
Comment (kurtisvg) [inline content/en/_index.md:153]: I don't think we should take in arguements. That goes against the "No CLIs" idea.

File: content/en/_index.md
Comment (kurtisvg) [inline content/en/_index.md:212]: I kinda like this - @grayside what do you think about adding these headers to the Java sample?

I'm a little worried folks might think the need to add this comments verbatim though. Maybe clarify in the comment or description for this bullet point?

File: content/en/_index.md
Comment (kurtisvg) [inline content/en/_index.md:307]: `exampleSnippet` should have this arg, "main" should not.

File: content/en/_index.md
Comment (kurtisvg) [inline content/en/_index.md:342]: Hmm this is interesting too - feels like we might want to be consistent with Java on this. @grayside thoughts?

File: content/en/_index.md
Comment (kurtisvg) [inline content/en/_index.md:394]: Does anything need to be done to "clean up" the client? Is it threadsafe?

**Review by kurtisvg — CHANGES_REQUESTED** [review]

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:307]: How do the tests pass in arguments to the snippet then?

(Would we have to add it to `exports`?)

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:141]: I'd argue region tags should **not** show the `main()` method, but I'm curious as to what other Node folks (@bcoe?) have to say.

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:394]: No - and Node is (from the programmer's standpoint) single-threaded.

File: content/en/_index.md
Comment (kurtisvg) [reply content/en/_index.md:307]: So I think if you are using this pattern:

```js

function main(){
      // ...
      exampleSample(arg1, arg2)
}

function exampleSample(arg1, arg2){
      // ...
}

if module.require == "main" {
     main()
}
```

Then the following should be true:
1. Main should only be a stub used by the user to set variables and run in whichever way they are most comfortable
2. testing can still happen on exampleSample directly, since the test can set args needed for it directly.

File: content/en/_index.md
Comment (kurtisvg) [reply content/en/_index.md:141]: The goal is that users need to be able to copy and paste the snippet directly from the website and run it from whatever environment they are using. The point of the `main` is to provide a single entrypoint and streamline their execution, even if they aren't using a fancy IDE - they can just execute from the command line after editing the file with VIM or whatever they are using.

File: content/en/_index.md
Comment (kurtisvg) [reply content/en/_index.md:394]: Can we update the comment to reflect that resources used by the client in the background are automatically cleaned up when it is garbage collected?

Is Node.JS _always_ single threaded? Python is _normally_ single threaded, but the ability to spawn different thread/processes exists. [It looks like Node offers something similar](https://nodejs.org/api/worker_threads.html)?

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:307]: I presume the _tests_ can `require(...)` a snippet file to invoke `main`?

(If so - how would the tests pass values to the sample itself? Or are the not supposed to?)

Comment (kurtisvg) [reply content/en/_index.md:307]: I'm not quite sure what you're asking, but here's an example of this pattern in java-land (with one slight derivation that this before we used "main" - everything else is the same):

Here's the "main" method: https://github.com/GoogleCloudPlatform/java-docs-samples/blob/c5df4758e5fbbb2261338e69fdfd1f8794a5f7d1/kms/src/main/java/kms/CreateKeyRing.java#L27

Here's the "exampleSnippet": https://github.com/GoogleCloudPlatform/java-docs-samples/blob/c5df4758e5fbbb2261338e69fdfd1f8794a5f7d1/kms/src/main/java/kms/CreateKeyRing.java#L36

Here's the test calling "exampleSnippet": https://github.com/GoogleCloudPlatform/java-docs-samples/blob/c5df4758e5fbbb2261338e69fdfd1f8794a5f7d1/kms/src/test/java/kms/SnippetsIT.java#L317

Comment (ace-n) [reply content/en/_index.md:307]: There *is* parameter-passing going on there. In Java, that's easy - it's a `public` class with a `public` method.

In Node, to make a method `public` you have to add it to `exports` - otherwise it's (effectively) `private`.

Comment (ace-n) [reply content/en/_index.md:307]: My only (minor) objection is wordiness - we'd have to do one of the following. Of the two, **Option B** feels cleaner to me.

**Option A**
```js
exports.exampleSnippet = (arg1, arg2) => {
   ...
}

function main() {
  arg1 = ...
  arg2 = ...

  // Notice the extra `exports` here
  exports.exampleSnippet(arg1, arg2);
}
```

**Option B**
```js
const exampleSnippet = (arg1, arg2) => {
  ...
}

function main() {
  arg1 = ...
  arg2 = ...

  // No `exports`! :)
  exampleSnippet(arg1, arg2)
}

exports.exampleSnippet = exampleSnippet;
```

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:212]: > I'm a little worried folks might think the need to add this comments verbatim though.

Added a paragraph about this before the samples themselves.

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:153]: Reorganized this - the `main` method only invokes snippets with **default** arguments now.

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:140]: I'm fine with `main`. Per our discussion in a different comment - replaced this with a no-arg `main` function and a call to `exports`.

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:141]: Point taken. I presume we should also include this part?

> if (require.main === module) {
>      // Run the snippet
>      main();
>  }

Comment (kurtisvg) [reply content/en/_index.md]: Yes, I think so.

File: content/en/_index.md
Comment (kurtisvg) [reply content/en/_index.md:221]: We chatted about this in the working group yesterday and the consensus was that we don't want these comments in the snippet for now.

File: content/en/_index.md
Comment (kurtisvg) [reply content/en/_index.md]: We also chatted about this one in the WG meeting, and the consensus was that we want to avoid showing the "wrong example" for now - there are some example of this in other readability guides, but they are able to show a red vs green example to more clearly mark which one was good and which one wasn't. 

So I think for now we only want the "right" way to do it showing.

File: content/en/_index.md
Comment (kurtisvg) [inline content/en/_index.md:294]: I think this is better understood with an actual example, like the Java example uses:

```suggestion
    function main() {
      // TODO(developer): Replace these variables before running the sample.
      const projectId = "my-project-id";
      const filePath = "path/to/image.png";
      exampleSnippet(arg1, arg2);

      exampleSnippet(arg1, arg2, ...);
    }
```

File: content/en/_index.md
Comment (kurtisvg) [inline content/en/_index.md:323]: It also seems like this should be consistent with the Java sample, and just show the `exampleSnippet` bit

File: content/en/_index.md
Comment (kurtisvg) [reply content/en/_index.md:402]: This looks unresolved still?

**Review by kurtisvg — CHANGES_REQUESTED** [review]

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:221]: OK, I removed them.

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md]: OK - done.

**Comment by kurtisvg** [comment]
Also, maybe tangentially, I noticed that "Node.js" tabs look like they are broken -when I click on one tab it doesn't switch all the Tabs, but it does for "Java". I switched the tabs to "Node" and the tab switching works, but the linting is lost. I switched to "Javascript" and both worked?

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:402]: I thought about this a bit more - usually, these objects are stored globally and destroyed when the program itself terminates. I'm not sure if they support programmatic garbage collection (as I've never needed to do that).

Is that an OK approach to recommend, or should I check to see if they're properly GC'ed (i.e. don't leak memory when going out of scope)?

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:103]: Removed.

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:294]: OK - done.

File: content/en/_index.md
Comment (kurtisvg) [reply content/en/_index.md:389]: The two things that I've seen come up in Java and Python are:
1. Is this thing threadsafe?
2. How do I clean up background resources?

I know that _normally_ we tend to recommend global usage, but there seem to almost certainly users that have questions about those 2 things, so some advice on how they should handle them would be good. 

Maybe Ben has some insights into both?

Comment (kurtisvg) [reply content/en/_index.md:389]: I think even if the answer is "These hang onto resources until the process ends", we try to say that as well.

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:315]: Node methods don't have explicit return types, so I'd argue we should _at least_ call out that we're not returning anything here.

(In Java, that's implied by `void`.)

File: content/en/_index.md
Comment (grayside) [inline content/en/_index.md:216]: I realize this use of `DLP`/`dlp` is a fairly common JS practice, but it has always bothered me. I'd be interested in learning more about whether this is a bad habit that gets repeated or something we actively encourage.

File: content/en/_index.md
Comment (grayside) [inline content/en/_index.md:356]: Are there still official clients that don't use ADC by default?

File: content/en/_index.md
Comment (kurtisvg) [reply content/en/_index.md:315]: Does calling the function without assignment imply that it's not returning a value or just that it's not being used? This isn't typically the case for most languages I'm familiar with. 

or were you thinking something like a comment in `exampleSnippet`? Maybe like `// Avoid using "return"... instead rely on testing output from "console.log("")"`

Comment (ace-n) [reply content/en/_index.md:315]: The latter - I added this as a comment:
> // (This snippet should not `return` anything.)

File: content/en/_index.md
Comment (grayside) [reply content/en/_index.md:389]: I think bcoe or alexander-fenster could help us clarify this.

File: content/en/_index.md
Comment (grayside) [inline content/en/_index.md:416]: Out of Scope: references to another doc should probably be followed by the URL.

File: content/en/_index.md
Comment (grayside) [inline content/en/_index.md:425]: In Node.js, I think log or write are used more commonly, as `print` isn't really in a stdlib term.

```suggestion
    // Unhandled exceptions will cause Node.js to crash. Log exception
    // messages via console.error(), and handle any non-fatal exceptions.
```

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:416]: BTW - I looked around [here](https://google.github.io/styleguide/javaguide.html#s6.2-caught-exceptions), and didn't find anything relevant. (Maybe this is an _Effective Java_-sort of thing?)

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:425]: Done.

File: content/en/_index.md
Comment (kurtisvg) [reply content/en/_index.md:416]: It might be an effective Java thing - I copied this from the Java samples style guide, so we probably have it wrong in there. I'd be okay with shortening it to "Always catch the most specific type of Exception, instead of a more general one." - I can also do that in a different PR if need be.

File: content/en/_index.md
Comment (bcoe) [inline content/en/_index.md:141]: With the existing sample guidelines, `main` is usually not part of the snippet, rather it looks something like this:

```js
function main(name) {
  // [START secretmanager_v1_generated_SecretManagerService_GetSecret_async]
  /**
   * TODO(developer): Uncomment these variables before running the sample.
   */
  /**
   *  Required. The resource name of the Secret google.cloud.secretmanager.v1.Secret, in the format `projects/* /secrets/*`.
   */
  // const name = 'abc123'

  // Imports the Secretmanager library
  const {SecretManagerServiceClient} = require('@google-cloud/secret-manager').v1;

  // Instantiates a client
  const secretmanagerClient = new SecretManagerServiceClient();

  async function callGetSecret() {
    // Construct request
    const request = {
      name,
    };

    // Run request
    const response = await secretmanagerClient.getSecret(request);
    console.log(response);
  }

  callGetSecret();
  // [END secretmanager_v1_generated_SecretManagerService_GetSecret_async]
}
```

☝️

**Review by bcoe — COMMENTED** [review]
The samples for Node.js in this PR seem to deviate a bit from the g3doc. If you decide to go in this direction this is fine, but you should coordinate with @sofisl and @amanda-tarafa who are working on sample generation, as generated samples are based on the current internal guide.

@sofisl is the better reviewer of this PR than me, as they've been working on the SnippetGen project 👍

File: content/en/_index.md
Comment (bcoe) [reply content/en/_index.md:389]: > Is Node.JS always single threaded?

Node.js doesn't have thread primitives like other programming languages, such that you would have to worry about contention on a global instance of the client.

There are [workers](https://nodejs.org/api/worker_threads.html#worker-threads) but they only communicate through a message passing API.

We recommend for environments like Cloud Functions that users create and cache a client.

> Does anything need to be done to "clean up" the client?

Nothing should need to be done to cleanup the client. Coming back to the Cloud Run or Cloud Function example, it's recommended you keep a cached version of the client in the global scope, to reduce startup overhead on subsequent requests. This client will not be GCd until the function shuts down.

If you did create a client on each request, it will be GCd as soon as its enclosing function is done executing.

Comment (bcoe) [reply content/en/_index.md:389]: > I think even if the answer is "These hang onto resources until the process ends", we try to say that as well.

It will hang on to resources in memory until it has no reference; if you create the client in the global scope, this will be process termination, if you create it in a closure, this will be when that closure finishes execution.

Comment (grayside) [reply content/en/_index.md:389]: Thanks Ben, sounds like the answer on "do we close/cleanup the client explicitly" is a "no".

For reusing clients across inbound requests, we skirt the edges of best practices in the guidelines, we could perhaps have a comment recommending the client object have global scope. I don't think we want to make it an explicit guideline for all samples.

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:416]: I went ahead and did it here. 👍

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md:141]: I applied @bcoe's suggestion here.

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md]: Used the following as a comment:
```
// Instantiates a reusable client object in the global scope.
// (Node.js will automatically clean it up when the script terminates.)
```

File: content/en/_index.md
Comment (kurtisvg) [reply content/en/_index.md:141]: This format (having a function at the top with the env variables) was the preferred format from the samples working group in 2019. Essentially we were trying to optimize for whatever makes it easier for the user to copy and paste the sample, and whatever way they saw fix.

From what I remember of the discussion, we had identified 3 main ways to run it:
1. User runs it from a heavier IDE (vscode, intellij, etc)
2. User calls it from some existing application or code
3. User executes it from the command line (vim, emacs, etc)

The reason we opted for a function was that because it covered use-cases 1 and 2 quite nicely - without a no-arged function there wasn't a super easy way to trigger it from the IDE. The only thing Java has changed from those discussions was to make the function "main" so that use-case 3 was more streamlined as well. (We also strongly wanted to avoid commented out variables and code because there were lots of examples of them being outdated and/or wrong since they bypassed most lint and compile checks). 

I don't have a particular preference to if the function is called "main", but it does seem like there's value to having a function that acts as an anchor for the user to right+click and run. Unless we have a strong inclination that Node developers will never do this, it does seem like it it's useful to show?

File: content/en/_index.md
Comment (kurtisvg) [reply content/en/_index.md:356]: FWIW, the Java healthcare samples don't: https://github.com/GoogleCloudPlatform/java-docs-samples/blob/306387ec973a6e97f733d602bd3c31c5dbab8cd7/healthcare/v1/src/main/java/snippets/healthcare/fhir/FhirStoreDelete.java#L57

File: content/en/_index.md
Comment (amanda-tarafa) [reply content/en/_index.md:141]: I think it's very useful for the general sample, handwritten ones in particular. For Phase 1 SnippetGen samples, just take into account the following:

- The samples are not runnable, as we are not able initialize the request properly. Users still have to make significant changes to the code. You can see the sample more as a template actually.
- The samples are not testable, this is because of volume of samples. The generators are thouroghly tested though.
- The samples will be hardly shown on handwritten docs. They are shown in generated docs, i.e. API and library reference docs, for which the less code the better (see an [example](https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Iam.Admin.V1/latest/Google.Cloud.Iam.Admin.V1.IAMClient#Google_Cloud_Iam_Admin_V1_IAMClient_CreateRole_Google_Cloud_Iam_Admin_V1_CreateRoleRequest_Google_Api_Gax_Grpc_CallSettings_)), as in this context users are not looking for a specific working sample but for an a quick look on how to call a given method.

But as I said, not hard rules for SnippetGen. I also think the general style guide should be more focused on handwritten samples, and on SnippetGen we can, on a language by language basis, make exceptions based on the above considerations. For SnippetGen Phase 2, where generated samples will be more handwritten-like we will have different aspects to consider.

**Comment by ace-n** [comment]
**N.B:** this is running into merge conflicts due (mostly) to whitespaces. 😬 

Once this PR is approved, I'll close it and create a clean version elsewhere (that we can rubber-stamp and merge).

**Review by sofisl — APPROVED** [review]

File: content/en/_index.md
Comment (sofisl) [inline content/en/_index.md:313]: nit: we usually don't use arrow functions.

Comment (ace-n) [reply content/en/_index.md]: Done (except for a lambda on line `239`).

File: content/en/_index.md
Comment (kurtisvg) [reply content/en/_index.md:139]: Reopening this since having the region tags here negates the purpose of this structure (that users can right click to run from an IDE). I tried out [outline in this comment](https://github.com/GoogleCloudPlatform/samples-style-guide/pull/28/files#r826420331) the thinking behind the structure that's proposed. I think maybe this sample as is doesn't work with the `exports.exampleSnippet = exampleSnippet;` at the bottom (unless JS has funkier scopes than I remember).

Is the hang up here on just `main()`? [In this PR](https://github.com/GoogleCloudPlatform/samples-style-guide/pull/41) I suggested I tried to clarify that it doesn't need to be _named_ main unless the language requires it. Would it be better to call it something like `runExampleSnippet` to make clear it's a stub for running `exampleSnippet`? Would that be more idiomatic for Node? cc @sofisl for her input as well

I'd think the pattern makes it really easy to run the samples and has a lot of value, but if you feel like it's at odds with being idiomatic for Node.js let me know.

File: content/en/_index.md
Comment (kurtisvg) [inline content/en/_index.md:306]: nit: I think this is supposed to have no args
```suggestion
    function main() {
```

File: content/en/_index.md
Comment (kurtisvg) [reply content/en/_index.md:216]: @ace-n @sofisl this comment is still also open - not sure if you folks have any preferences

**Review by kurtisvg — COMMENTED** [review]
it would be nice if we could rebase this off of main - would make it a bit easier to compare diffs.

**Review by kurtisvg — COMMENTED** [review]
LGTM: I think there are two small issues to resolve:
- https://github.com/GoogleCloudPlatform/samples-style-guide/pull/28/files/1f94f25df8bad0bab78ec6be6ff360b9c003250b#r829647134
- https://github.com/GoogleCloudPlatform/samples-style-guide/pull/28/files/1f94f25df8bad0bab78ec6be6ff360b9c003250b#r824220103

**Review by kurtisvg — APPROVED** [review]
LGTM: I think there are two small issues to resolve:
- https://github.com/GoogleCloudPlatform/samples-style-guide/pull/28/files/1f94f25df8bad0bab78ec6be6ff360b9c003250b#r829647134
- https://github.com/GoogleCloudPlatform/samples-style-guide/pull/28/files/1f94f25df8bad0bab78ec6be6ff360b9c003250b#r824220103

File: content/en/_index.md
Comment (sofisl) [reply content/en/_index.md:216]: Hm, I'm not sure I understand the concern. This is how we have it currently in samples: https://github.com/googleapis/nodejs-dlp/blob/bd3d585a773571bc5eb243f34f9891c4b16151d7/samples/categoricalRiskAnalysis.js#L37.

I would prefer going with custom, but happy to hear out the concern further if I'm not understanding it correctly.

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md]: Done.

File: content/en/_index.md
Comment (grayside) [reply content/en/_index.md:216]: The idiom of `DLP` and `dlp` as two different variables bothers me. Is this a thing that's often done, or a thing we want people to do? I personally find this a poor UX, so I'm using this moment to briefly question the status quo.

Comment (ace-n) [reply content/en/_index.md:216]: A few ideas: would it be better to:
- label these variables as `library`/`client`?
- drop the library import, and directly initialize the client?

Of these, I feel like the first (`library`/`client`) is the easiest to understand.

Comment (sofisl) [reply content/en/_index.md]: These variables should be labeled as library/client DLP/dlp, respectively.

Comment (ace-n) [reply content/en/_index.md]: @sofisl to clarify - are you endorsing the following?

```
const library = require(...)

const client = library()
```

**Comment by ace-n** [comment]
@sofisl and I chatted about the `DLP`/`dlp` issue. We're reluctant to rename `DlpServiceClient` (because it's `export`ed under that name), but the _initialized_ client object could be named something like `dlpClientInstance`.

@grayside does this address your concerns?

**Review by bcoe — APPROVED** [review]
👍 just so I don't have a yellow "awaiting review" icon next to my name.

I'm supportive of whatever folks land on with @sofisl.

File: content/en/_index.md
Comment (grayside) [reply content/en/_index.md]: I'm not asking that we change this, I'm asking if this common Node.js language practice is something we recommend. if so, this is fine. If not, then it seems like we should change it.

Comment (sofisl) [reply content/en/_index.md]: @grayside, sorry for misunderstanding, yes this is what we currently recommend in our guidelines, insofar as we don't recommend renaming the library that is exported (e.g., `const {thing} = require('import')`, vs `const {thing} as anotherThing = require('import')`

Personally, I think renaming the export is more confusing if someone wanted to go digging into source code. After discussing offline with Ace and understanding the context a bit better, I'd rather us rename the client to something less confusing rather than the library:

```
  // Imports the Dlp library
  const {DlpServiceClient} = require('@google-cloud/dlp').v2;

  // Instantiates a client
  const dlpClientInstance = new DlpServiceClient();
```

Comment (grayside) [reply content/en/_index.md]: I'm comfortable with whatever we believe is idiomatic and good UX for Node.js developers, whether that's the existing approach I personally find confusing, or renaming the client.

I understand in python they often use `client` or (to disambiguate) `dlp_client`, so your proposal makes sense to me.

**Review by grayside — APPROVED** [review]

**Comment by ace-n** [comment]
@sofisl over to you for one final (double-check-y) LGTM. 🙂 

**Please merge for me** once you've approved - thanks!

**Review by sofisl — APPROVED** [review]

File: content/en/_index.md
Comment (sofisl) [inline content/en/_index.md:247]: Sounds like we landed on something different, I'd make sure we update to: 
```
const {DlpServiceClient} = require('@google-cloud/dlp').v2;




  // Instantiates a client

  const dlpClientInstance = new DlpServiceClient();
```

And it should probably be changed in line 346 as well.

**Review by sofisl — APPROVED** [review]
LGTM when final comment is addressed.

File: content/en/_index.md
Comment (ace-n) [reply content/en/_index.md]: Done.

**Pull request merged**
